### PR TITLE
V160 and boundary

### PR DESCRIPTION
**Pull request by SimonDarksideJ**
# XRTK - Mixed Reality Toolkit Pull Request

## Overview

Oculus API update to latest and new Boundary implementation

## Changes

* Updated the Oculus runtime to V1.60 (V28)
* Added the boundary implementation for Oculus

### CODE REVIEW

File: XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        {
+            base.Enable();
+
+            boundarySystem.SetupBoundary(this);
+        }
+
+        #region IMixedRealityBoundaryDataProvider Implementation
+
+        /// <inheritdoc />
+        public BoundaryVisibility Visibility => BoundaryVisibility.Visible;
```
Comment (FejZa) [inline XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs:42]: If there is no API to query the current visibility state for boundary this should be "Unknown".

**Review by FejZa — CHANGES_REQUESTED** [review]

**Comment by FejZa** [comment]
Looks good. Just one little change for the visibility property.

File: XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs (hunk 1 above)
Comment (SimonDarksideJ) [reply XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs:42]: Can you clarify what the purpose of testing "visibility" is?  Couldn't see any comment documentation to state what it expects from the API.
Is it:

* Is the boundary currently visible?
* Is there a boundary to be visible?
* Has a boundary been configured?

Just so I know what to hook it up to.

Comment (SimonDarksideJ) [reply XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs]: Implementation updated

File: XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                }
+            }
+
+            return new Vector3[0];
+        }
+
+        /// <summary>
+        /// Returns a vector that indicates the spatial dimensions of the specified boundary type. (x = width, y = height, z = depth)
+        /// </summary>
+        public Vector3 GetDimensions(OculusApi.BoundaryType boundaryType)
```
Comment (StephenHodgson) [inline XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs:125]: @SimonDarksideJ unused?

File: XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+        private static int cachedVector3fSize = Marshal.SizeOf(typeof(OculusApi.Vector3f));
+        private static OculusApi.OVRNativeBuffer cachedGeometryNativeBuffer = new OculusApi.OVRNativeBuffer(0);
+        private static float[] cachedGeometryManagedBuffer = new float[0];
+
+        /// <summary>
+        /// Returns an array of 3d points (in clockwise order) that define the specified boundary type.
+        /// All points are returned in local tracking space shared by tracked nodes and accessible through OVRCameraRig's trackingSpace anchor.
+        /// </summary>
+        public Vector3[] GetGeometry(OculusApi.BoundaryType boundaryType)
```
Comment (StephenHodgson) [inline XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs:79]: If the method isn't part of the interface it should be private

File: XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs (hunk 2 above)
Comment (SimonDarksideJ) [reply XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs:125]: I can remove it, but I retained it for future use.

File: XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+            geometry.AddRange(oculusGeometry);
+            return true;
+        }
+
+        #endregion  IMixedRealityBoundaryDataProvider Implementation
+
+        #region Oculus Utils
+
+        private static int cachedVector3fSize = Marshal.SizeOf(typeof(OculusApi.Vector3f));
```
Comment (StephenHodgson) [inline XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs:71]: can be readonly

File: XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        private static OculusApi.OVRNativeBuffer cachedGeometryNativeBuffer = new OculusApi.OVRNativeBuffer(0);
```
[the remark this answers is not in the document]
Comment (SimonDarksideJ) [reply XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs:72]: Done

File: XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                                z = cachedGeometryManagedBuffer[3 * i + 2],
+                            }.FromFlippedZVector3f();
+                        }
+
+                        return points;
+                    }
+                }
+            }
+
+            return new Vector3[0];
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply XRTK.Oculus/Packages/com.xrtk.oculus/Runtime/Providers/BoundarySystem/OculusBoundaryDataProvider.cs:127]: (if it never gets updated then it'll be the same thing anyway)

**Review by StephenHodgson — APPROVED** [review]

**Pull request merged**